### PR TITLE
Add RTK lat/lon/alt High Precision components to SensorGps.msg for centimeter level GPS accuracy 

### PR DESCRIPTION
**Pull request by slgrobotics**
This PR seeks to match [https://github.com/PX4/PX4-GPSDrivers/pull/136](https://github.com/PX4/PX4-GPSDrivers/pull/136) - adding high precision RTK lat/lon/alt components from MSG_NAV_HPPOSLLH

### Solved Problem
This creates a possibility to operate coordinates with precision necessary to drive land based Rovers, for example, agricultural machines.

### Solution
please follow the link above for details

### Note:
once 1e-9 precision is available in _gps_position ORB message, it can be used in the following modules, as far as I know:
- EKF2
- GPS Blending

For my Rover project, I modified RoverPositionControl.cpp to subscribe to ORB_ID(sensor_gps) and use precise lat/lon for course planning. This won't be necessary if EKF2 could deliver centimeter level precision coordinates.

### CODE REVIEW

**Comment by slgrobotics** [comment]
I do agree that converting the *lat, lon, alt* and *alt_ellipsoid* fields in *SensorGps.msg* makes the most sense. From what I see in the code, degrees in WGS84 for lat/lon and meters for alt seems to be a standard.

To feel the extent of the changes I did a quick experiment - renamed the four fields (adding a *_dbl* postfix and making them float64) and went through all files which broke the compile (I used *"make emlid_navio2"* and *"make px4_sitl gazebo-classic_r1_rover"*). I just hacked the code to compile, not even trying to make calculations correct. 

Here is the list of files I had to modify in the process:
```
xxx:~/src11/PX4-Autopilot$ git status
On branch RTK-HPPOSLLH
Changes not staged for commit:
	modified:   msg/SensorGps.msg
	modified:   src/drivers/gps/devices (modified content)
	modified:   src/drivers/osd/msp_osd/uorb_to_msp.cpp
	modified:   src/drivers/rc_input/crsf_telemetry.cpp
	modified:   src/drivers/rc_input/ghst_telemetry.cpp
	modified:   src/examples/fake_gps/FakeGps.cpp
	modified:   src/examples/fake_magnetometer/FakeMagnetometer.cpp
	modified:   src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp
	modified:   src/modules/commander/HomePosition.cpp
	modified:   src/modules/commander/baro_calibration.cpp
	modified:   src/modules/commander/mag_calibration.cpp
	modified:   src/modules/ekf2/EKF2.cpp
	modified:   src/modules/local_position_estimator/sensors/gps.cpp
	modified:   src/modules/mavlink/mavlink_receiver.cpp
	modified:   src/modules/mavlink/streams/GPS2_RAW.hpp
	modified:   src/modules/mavlink/streams/GPS_RAW_INT.hpp
	modified:   src/modules/mavlink/streams/OPEN_DRONE_ID_LOCATION.hpp
	modified:   src/modules/mavlink/streams/UAVIONIX_ADSB_OUT_DYNAMIC.hpp
	modified:   src/modules/navigator/geofence.cpp
	modified:   src/modules/sensors/vehicle_gps_position/gps_blending.cpp
	modified:   src/modules/simulation/sensor_gps_sim/SensorGpsSim.cpp
	modified:   src/modules/simulation/simulator_mavlink/SimulatorMavlink.cpp
```
From what I saw so far, the EKF2 has its own *struct gpsMessage* (...PX4-Autopilot/src/modules/ekf2/EKF/common.h:158) with int32_t members - this might lead to deeper dive in the rabbit hole. Mavlink is a bit tricky too. Other areas seem to be rather trivial. 

In any case, this is way above my head - while I can prepare a draft PR (changing the files above to have correct conversions), I cannot test it to full extent. It will also require coordination with at least GPS and EKF2 submodules maintainers.

Let me know how I can help.

**Comment by slgrobotics** [comment]
Thanks, @dagar  - I'll start on the PR ASAP. I just need a decision from you - on the naming and on what we do with the altitudes.
1. Naming. While *latitude* and *longitude* are nice and natural, there's 380 matches in the codebase for the first and 371 for the second. Making a more unique name has a benefit of immediately finding all occurrences of a variable by a simple global search.
2. Altitudes (*alt* and *alt_ellipsoid*) as they are - are already in millimeters, RTK hp components just add one decimal digit - 0.1mm granularity - totally unnecessary. Consumers do convert the int32 to float, yielding meters. We can leave these altitudes untouched and work with lat/lon only - to avoid injecting unnecessary bugs into the code, or can go ahead with submillimeter float64's for diligence. In either case, while we are at it, I would suggest renaming *alt* to *alt_msl* for easier search. We can also use full *altitude_msl* and *altitude_ellipsoid* names.
3. Usually in my code I append measurement units to variables that represent physical phenomena. This way the variables discussed would be *latitude_deg*, *longitude_deg*, *altitude_msl_m* and *altitude_ellipsoid_m* (*altitude_msl_mm* and *altitude_ellipsoid_mm* if left as int32 and simply renamed). Just a thought.

Let me know what you've decided, I am fine with any choice.

**Comment by slgrobotics** [comment]
@dagar  - please review the above, I need your input on naming and alt handling to proceed. Thanks!

**Comment by slgrobotics** [comment]
@dagar  - I checked in all changes related to renaming *lon/lat/alt*

It compiles fine under *"make emlid_navio2"* and *"make px4_sitl gazebo-classic_r1_rover"*

Related GPSDriver changes are now in new PR:  https://github.com/PX4/PX4-GPSDrivers/pull/136

I wasn't able to compile with UAVCAN enabled, so there are places where verification and more changes might be needed:
```
src/drivers/uavcan/sensors/gnss.cpp : 340
src/drivers/uavcannode/Publishers/GnssFix2.hpp : 84
src/drivers/cyphal/Publishers/udral/Gnss.hpp : 69
```
I am pretty sure there are some more places that need a little touch, and, of course, testing.
I will be doing testing on my Rovers, but my vehicle park is very limited.

**Comment by slgrobotics** [comment]
I will go through "checks" and fix the errors where possible.

**Comment by slgrobotics** [comment]
@dagar - here is the status:

1. I temporarily changed .gitmodules to point to RTK-HPPOSLLH branch of https://github.com/slgrobotics/PX4-GPSDrivers, to allow builds
2. I corrected build errors, including astyle format ones
3. Jenkins seems to build everything fine, but chokes on "make px4_fmu-v5_test". The code doesn't fit in FLASH_AXIM. Same happend when I build on my machine, but "make px4_fmu-v5" finishes fine.
4. Not sure why "tests" phase of "make px4_fmu-v4_test" has failed - it builds fine on my machine at least.

Let me know what else can I do.

**Comment by slgrobotics** [comment]
I fixed errors related to double/float promotion in CLANG (make clang-tidy-quiet) and all other checks.

"make tests-coverage" succeeded on my Ubuntu 20.04 PC, all 138 tests passed, created coverage/lcov.info (11MB file)

I'd guess we need another full build done.

File: src/drivers/osd/msp_osd/uorb_to_msp.cpp
Comment (bkueng) [inline src/drivers/osd/msp_osd/uorb_to_msp.cpp:319]:
```suggestion
		raw_gps.alt = static_cast<int16_t>(vehicle_gps_position.altitude_msl_m*100);
```

File: src/drivers/transponder/sagetech_mxs/SagetechMXS.cpp
Comment (bkueng) [inline src/drivers/transponder/sagetech_mxs/SagetechMXS.cpp:495]:
```suggestion
	const bool my_loc_is_zero = (fabs(_gps.latitude_deg) < FLT_EPSILON) && (fabs(_gps.longitude_deg) < FLT_EPSILON);
```

File: src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp
Comment (bkueng) [inline src/modules/attitude_estimator_q/attitude_estimator_q_main.cpp:223]:
```suggestion
						       (float)gps.longitude_deg));
```

**Review by bkueng — COMMENTED** [review]
Thanks for going through this.

Can you increase the data version to 2 in https://github.com/PX4/PX4-Autopilot/blob/dccfcb85c13180e6f8a64ef9dbcd75627933ff8f/src/modules/logger/logger.cpp#L2106 and add a comment regarding the gps message update?

And can you provide a log after all the updates (using the rover is fine)?
I can then update flight review accordingly.

File: src/drivers/osd/msp_osd/uorb_to_msp.cpp
Comment (slgrobotics) [reply src/drivers/osd/msp_osd/uorb_to_msp.cpp:319]: I am working on the rest of the changes (accepting all) - but this one needs attention. I think original division by 10 (mm->cm conversion) was conflicting with the units specifies in the _msp_raw_gps_t_  message (meters):

src/drivers/osd/msp_osd/uorb_to_msp.cpp : 313
```
	// initialize result
	msp_raw_gps_t raw_gps {0};

	if (vehicle_gps_position.fix_type >= 2) {
		raw_gps.lat = static_cast<int32_t>(vehicle_gps_position.latitude_deg * 1e7);
		raw_gps.lon = static_cast<int32_t>(vehicle_gps_position.longitude_deg * 1e7);
		raw_gps.alt = static_cast<int16_t>(vehicle_gps_position.altitude_msl_m);
```
src/drivers/osd/msp_osd/msp_defines.h : 383
```
// MSP_RAW_GPS reply
struct msp_raw_gps_t {
	uint8_t  fixType;       // MSP_GPS_NO_FIX, MSP_GPS_FIX_2D, MSP_GPS_FIX_3D
	uint8_t  numSat;
	int32_t  lat;           // 1 / 10000000 deg
	int32_t  lon;           // 1 / 10000000 deg
	int16_t  alt;           // meters
	int16_t  groundSpeed;   // cm/s
	int16_t  groundCourse;  // unit: degree x 100, centidegrees
	uint16_t hdop;
} __attribute__((packed));
```
What is your decision on this? is it meters or centimeters that we should pass to msp?

**Comment by slgrobotics** [comment]
@bkueng  - I uploaded log file from my full size gas-powered zero-turn lawnmower:

https://review.px4.io/plot_app?log=6c95bdf7-57a8-4e1b-b292-1d2e1d4362a7

I fixed all I could, need your decision on src/drivers/osd/msp_osd/uorb_to_msp.cpp

Thanks!

File: src/drivers/osd/msp_osd/uorb_to_msp.cpp
Comment (bkueng) [reply src/drivers/osd/msp_osd/uorb_to_msp.cpp:319]: Hmm it looks like the specification changed: https://github.com/betaflight/betaflight/blob/master/src/main/msp/msp.c#L1522.
Probably best to use the new one then, and update the comment in the struct.

**Comment by bkueng** [comment]
Thanks @slgrobotics. Does this include my comments from the GPS submodule as well?

File: src/drivers/osd/msp_osd/uorb_to_msp.cpp
Comment (slgrobotics) [reply src/drivers/osd/msp_osd/uorb_to_msp.cpp:319]: Understood, will update comment in _msp_defines.h_ ("meters"->"centimeters since MSP API 1.39, meters before") and accept the "/100" in _uorb_to_msp.cpp_

**Comment by slgrobotics** [comment]
@bkueng  - Updated as you suggested, including _GPSDrivers_. Let me know if I understood your suggestions correctly. Thanks!

**Comment by slgrobotics** [comment]
Just in case - a fresh log here:

 https://review.px4.io/plot_app?log=b835e623-eb13-4720-b964-a3562e1d7ce5

**Comment by bkueng** [comment]
PR for flight review: https://github.com/PX4/flight_review/pull/263

File: .gitmodules
Comment (bkueng) [inline .gitmodules:20]: Can you rebase, squash the commits, update the gps submodule and revert this? Then it should be good to merge.

**Comment by slgrobotics** [comment]
I updated the gps submodule and reverted .gpsmodules, but when it came to rebasing and squashing - things got messy.

I am relatively new to  github, and this is my first attempt on squashing (_rebase -i  RTK-HPPOSLLH~35_) - so, when I got through it I found many strangely squashed commits.

Here I had to do _"git reset --hard e3ba98c145"_ and push -f, to get back to a nice state (i.e. .gitmodules updated and then 
e4349a3 merged latest from main. _"This branch is 36 commits ahead of PX4:main."_).

Will try to squash again, but just in case - invited @bkueng as collaborator with, I hope, full rights to this branch.

**Comment by slgrobotics** [comment]
@bkueng   -- Ok, squashing seems to be "above my pay grade" ;-)

I tried to squash only my changes, and removed lines with others - here's what I've got on a guinea pig branch: 

https://github.com/slgrobotics/PX4-Autopilot/tree/RTK-HPPOSLLH-2

This was result of the following:
```
reword 0be6d770f0 Update SensorGps.msg, lat/lon/alt as double for RTK precision
squash 67b0c6600b Did the renaming, altitudes need work
squash a4ab470219 Everything replaced/corrected and compiles fine
squash 95970a64f9 Update SensorGps.msg
squash 8284936e24 More renaming and scaling
squash 9d8ab5301a Applied "make format" to pass "make check_format" test
squash 700fba1bdd Update .gitmodules
squash 10566d8a29 Submodule updated
squash cb34c56c00 Update sPort_data.cpp
squash 7841d2f567 Updated submodule GPSDrivers
squash 698b3968f0 Fixed more Jenkins build errors
squash c5937e6ed3 Apply suggestions from code review
squash cdbcbe7a0e Update logger.cpp
squash 3ed5a8d654 Update SagetechMXS.cpp
squash f7058b7a1b Updated per @bkueng suggestions
squash 8ed713385a Update devices
squash 365a305183 Update devices
squash e3ba98c145 Updated .gitmodules to use "main" branch of GPSDevices
```

I can force push of the same to this branch (RTK-HPPOSLLH) - just feel uncomfortable with submodules changes and possible difficulty in merging. The branch in its current state can be easily merged though.

Basically I need either a direction (what exactly the purpose of the squashing - to squash only my changes and ignore others, or to bring everything up in sync with main and then apply my changes in one squash?) - or I'd give a more qualified person full access to the RTK-HPPOSLLH branch to do remaining conversions.

**Review by bkueng — APPROVED** [review]
Apart from the submodule changes that slipped in your rebase looks correct. I did it now for your branch.
We generally want to make sure every commit compiles, and having a clean history while splitting independent changes into separate commits. As in this case everything is related, it makes sense to keep it all in one.

Let's wait for CI, otherwise it's good to merge.

**Comment by slgrobotics** [comment]
I just cloned the _"main"_  and my RTK-HPPOSLLH locally and compared the trees. I see my changes and some changes related to latest main check-ins (like UWB stuff) - but nothing in the submodules (rather than _GPSDrivers_ of course). So, I'd guess, the squashed RTK-HPPOSLLH branch is as good as it gets for the merge.

**Pull request merged**
